Enable Token Authorization by default

udaij12 opened the pull request:
## Description
Setting token authorization by default #3157

Users can disable using environment variables, command line, or with config.properties. 
Check Documentation for examples 


## Tests
Updating all existing tests to disable token authorization except for token authorization tests in pytest

namannandan commented on docs/token_authorization_api.md:
```diff
@@ -1,11 +1,14 @@
-2. Enable token authorization by adding the `--plugins-path /path/to/the/jar/files` flag at start up with the path leading to the downloaded jar files.
+Torchserve now supports token authorization by default.
+
+## How to disable Token Authorization
+1. Set global environment variable `TS_DISABLE_TOKEN_AUTHORIZATION=true`
+2. Add `--disable-token` to command line when running TorchServe.
+3. Add `disable_token_authorization=true` to config.properties file
```
Should we clarify that any one of these options can be used to disable token authorization?
```suggestion
## How to disable Token Authorization
1. Set global environment variable `TS_DISABLE_TOKEN_AUTHORIZATION=true` or
2. Add `--disable-token` to command line when running TorchServe or
3. Add `disable_token_authorization=true` to config.properties file
```

Also for `Set global environment variable TS_DISABLE_TOKEN_AUTHORIZATION=true` I think it may be good to link to https://github.com/pytorch/serve/blob/master/docs/configuration.md and mention that `enable_envvars_config=true` must be set to ensure that the environment variable takes effect.

namannandan commented on docs/token_authorization_api.md:
```diff
@@ -45,9 +48,7 @@
 ## Customization
 Torchserve offers various ways to customize the token authorization to allow owners to reach the desired result.
```
Could we double check that this statement on customizability is still valid. In case this is not intended, we can delete it or test using a plugin to see if `token` implementation provided in the frontend can be overridden by a plugin.

udaij12 replied:
Will remove customization section and move time to expiration to notes.  Plugin implementation no longer works

lxning reviewed (commented):
Why not enable the token authentication plugin by default to avoid big code changes and  copying/pasting Token class?

udaij12 commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -44,36 +60,46 @@ public void handleRequest(
+            }
+            chain.handleRequest(ctx, req, decoder, segments);
         }
-        chain.handleRequest(ctx, req, decoder, segments);
```
Noted

udaij12 commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -44,36 +60,46 @@ public void handleRequest(
-                | ClassNotFoundException e) {
+        } catch (Exception e) {
             e.printStackTrace();
             logger.error("TOKEN CLASS IMPORTED UNSUCCESSFULLY");
```
Kept it so that when TorchServe starts users can see if the keyfile was created successfully if enabled.

udaij12 commented:
> Why not enable the token authentication plugin by default to avoid big code changes and copying/pasting Token class?

I feel that if we are implementing token authorization as default it makes sense to integrate it into the frontend rather than as an external plugin as there is no need for that. It would also make future changes painful since we would have to build the plugin jar and publish to maven. It makes sense to integrate and remove the plugin completely so rather then duplication of code its just moving.

lxning reviewed (commented):
Can you add test for the priority b/w config file vs cmd, especially for the negative test cases (eg. conflict setting)?

udaij12 commented:
> Can you add test for the priority b/w config file vs cmd, especially for the negative test cases (eg. conflict setting)?

Added the test and added more clarification in the docs.

agunapal commented on docs/token_authorization_api.md:
```diff
@@ -1,11 +1,28 @@
-## Setup
-1. Download the jar files from [Maven](https://mvnrepository.com/artifact/org.pytorch/torchserve-endpoint-plugin) 
-2. Enable token authorization by adding the `--plugins-path /path/to/the/jar/files` flag at start up with the path leading to the downloaded jar files.
+Torchserve now supports token authorization by default.
```
@udaij12 Please create an RFC for this if not already done, before the PR lands

udaij12 replied:
What is an RFC Im not aware of what that is.

agunapal replied:
Example: https://github.com/pytorch/serve/issues/3066

agunapal commented on docs/token_authorization_api.md:
```diff
@@ -41,14 +58,7 @@
 ## Notes
 1. DO NOT MODIFY THE KEY FILE. Modifying the key file might impact reading and writing to the file thus preventing new keys from properly being displayed in the file.
-2. 3 tokens allow the owner with the most flexibility in use and enables them to adapt the tokens to their use. Owners of the server can provide users with the inference token if users should not mess with models. The owner can also provide owners with the management key if owners want users to add and remove models.
+2. Time to expiration is set to default at 60 minutes but can be changed in the config.properties by adding `token_expiration_min`. Ex:`token_expiration_min=30`
```
How is this going to work in a production setting? What would be the workflow?

udaij12 replied:
The workflow remains the same for production settings. Customers can disable token authorization to keep things the same or can change the token expiration depending on their use case. If a token is expired users have to reset the token and then use the updated token for future API calls.

agunapal commented on docs/token_authorization_api.md:
```diff
@@ -1,11 +1,28 @@
-## Setup
-1. Download the jar files from [Maven](https://mvnrepository.com/artifact/org.pytorch/torchserve-endpoint-plugin) 
-2. Enable token authorization by adding the `--plugins-path /path/to/the/jar/files` flag at start up with the path leading to the downloaded jar files.
+Torchserve now supports token authorization by default.
```
I suggest we specify this in multiple places: docker, kubernetes, kserve etc.
Its better to over-communicate this basic behavior change.

udaij12 replied:
Agreed. Was planning on creating a PR of documentation updates with both security related changes to inform of new behavior to TorchServe.

agunapal reviewed (approved):
LGTM. Please create an RFC like https://github.com/pytorch/serve/issues/3066 before merging

udaij12 commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -44,36 +60,44 @@ public void handleRequest(
             } else if (tokenType == TokenType.INFERENCE) {
                 checkTokenAuthorization(req, "inference");
             }
+        } else {
```
This is to prevent the token api from being called when token authorization is disabled.

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+    private static String inferenceKey;
+    private static Instant managementExpirationTimeMinutes;
+    private static Instant inferenceExpirationTimeMinutes;
+    private static Double timeToExpirationMinutes;
```
Lets get rid if the double bookkeeping here and clean out the code. Token is now part of TS so it can access ConfigManager. We can just read the time from there to have a single source of truth.

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -24,7 +40,7 @@ public class TokenAuthorizationHandler extends HttpRequestHandlerChain {
-    private static Class<?> tokenClass;
+    private static Token tokenClass;
     private static Object tokenObject;
     private static Double timeToExpirationMinutes = 60.0;
```
See comment on ```class Token``` and lets assume source of through to be config manager

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+    private String fileName = "key_file.json";
+    private String filePath = "";
+
+    public String updateKeyFile(FullHttpRequest req) throws IOException {
```
Lets not pull HttpRequest into this function but instead call generateKeyFile directly and parse request outside of Token. This class is highly critical, so we should keep it as lean as possible.

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+    }
+
+    // parses query and either returns management/inference or a wrong type error
+    public String parseQuery(FullHttpRequest req) {
```
Same here. This should be part of the handler not token.

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+    }
+
+    public void setFilePath(String path) {
+        filePath = path;
```
Lets read this directly from the config as well.

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+    }
+
+    // checks the token provided in the http with the saved keys depening on parameters
+    public boolean checkTokenAuthorization(FullHttpRequest req, String type) {
```
Same, lets move this up to handler and only get type + token as parameter

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+        return apiKey;
+    }
+
+    public Instant getInferenceExpirationTime() {
```
Unused?

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+        return inferenceExpirationTimeMinutes;
+    }
+
+    public Instant getManagementExpirationTime() {
```
Unused?

mreso commented on frontend/server/src/main/java/org/pytorch/serve/http/TokenAuthorizationHandler.java:
```diff
@@ -84,20 +108,208 @@ public static void setupTokenClass() {
+        return managementExpirationTimeMinutes;
+    }
+
+    public void setTime(Double time) {
```
See above. Use config manager.

mreso commented on frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:
```diff
@@ -46,6 +46,7 @@
 import org.apache.commons.cli.Options;
 import org.apache.commons.io.IOUtils;
 import org.pytorch.serve.archive.model.Manifest;
+import org.pytorch.serve.http.TokenAuthorizationHandler;
```
See above

mreso commented on frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:
```diff
@@ -438,6 +445,14 @@ public boolean isOpenInferenceProtocol() {
         return Boolean.parseBoolean(prop.getProperty(TS_OPEN_INFERENCE_PROTOCOL, "false"));
     }
 
+    public boolean setupToken() {
```
See above

mreso commented on frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:
```diff
@@ -1203,6 +1230,10 @@ public String getWorkflowStore() {
             return workflowStore;
         }
 
+        public String isTokenDisabled() {
```
Why do we use a string here instead of boolean?

udaij12 replied:
Using this function to set prop and prop takes strings not booleans.

mreso commented on test/pytest/test_token_authorization.py:
```diff
@@ -11,26 +9,9 @@
-            os.path.join(jar_path, jar_file), os.path.join(new_folder_path, jar_file)
-        )
-    os.chdir(REPO_ROOT)
+REPO_ROOT = os.path.dirname(os.path.abspath(__file__))
```
nit: Better to rename as its not repo root anymore.

mreso commented on test/pytest/test_utils.py:
```diff
@@ -59,6 +59,7 @@ def start_torchserve(
     no_config_snapshots=False,
     gen_mar=True,
     plugin_folder=None,
+    token=False,
```
lets make this less an a logic puzzle and name is disable_token so we do not need to invert below.

mreso commented on test/pytest/test_token_authorization.py:
```diff
@@ -215,3 +193,25 @@ def test_token_expiration_time(setup_torchserve_expiration):
+
+    test_utils.stop_torchserve()
+
+    assert response.status_code == 200, "Token check failed"
```
These is a test missing for the env variable and the different priorities with it.

mreso commented on ts_scripts/tsutils.py:
```diff
@@ -48,6 +48,7 @@ def start_torchserve(
     config_file="",
     log_file="",
     gen_mar=True,
+    token=False,
```
Same here, change to disable_token

mreso reviewed (changes requested):
The tests are not sufficient. We need to add disable through env variable and the different priorities. We also need to test updating the expired and not expired tokens (of both types).
I've left a bunch of other comments which are more related to a refactoring as we're e.g. introducing dependency cycles here. I'll create a follow up issue where we need to address these on a short timeline. We can ignore these for now as this is time sensitive but the testing needs to be improved before we can merge this.

udaij12 commented:
Added new tests 1. generating new tokens once expired 2. priority with env, cmd, and config properties 3. priority with env and cmd

mreso reviewed (approved):
Lets go ahead with this with the expectation that https://github.com/pytorch/serve/issues/3185 is addressed in the short-term.
